File: 41-First-Missing-Positive/go/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println("Hello World")
    nums := []int{
        7,8,9,11,12,
    }
    fmt.Println(firstMissingPositive(nums))
}

func firstMissingPositive(nums []int) int {
	n := len(nums)

	for i:=0; i< n; i++ {
		if (nums[i] >= 1 &&
			nums[i] <= n &&
			nums[i] != nums[nums[i]-1]) {
			nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
			i -= 1
		}
	}

	for i := range n {
		if nums[i] != i+1 {
			return i + 1
		}
	}
	return n + 1
}

File: 41-First-Missing-Positive/ts/main.ts
function firstMissingPositive(nums: number[]): number {
    let n = nums.length;
    
    for (let i = 0; i < n; i++) {
        if (nums[i] > 0 && nums[i] <= n && nums[i] !== nums[nums[i] - 1]) {
            nums[i], nums[nums[i] - 1] = nums[nums[i] - 1], nums[i];
            i--;
        }
    }
    
    for (let i = 0; i < n; i++) {
        if (nums[i] !== i + 1) {
            return i + 1;
        }
    }
    
    return n + 1;
}
